### ISSUE TITLE
use sparse arrays from xarray

### ISSUE THREAD

**Issue by FabianHofmann** [PyPSA/atlite#58]
As discussed with @euronion xarray supports sparse arrays since v0.14 see [the release notes](http://xarray.pydata.org/en/stable/whats-new.html#v0-14-1-19-nov-2019). 
It works pretty decent though the functionality is not well documented. 

For its use the [sparse package has to be installed](https://sparse.pydata.org/en/latest/install.html)

The two functions I use the [netallocation package](https://github.com/FRESNA/netallocation) are the following 

```python
import xarray as xr
from sparse import as_coo, COO


def as_sparse(ds):
    """
    Convert dense dataset/dataarray into a sparse dataset.

    Parameters
    ----------
    ds : xr.Dataset or xr.DataArray

    Returns
    -------
    xr.Dataset or xr.DataArray
        Dataset or DataArray with sparse data.

    """
    func = lambda data: COO(data) if not isinstance(data, COO) else data
    return xr.apply_ufunc(func, ds)


def as_dense(ds):
    """
    Convert sparse dataset/dataarray into a dense dataset.

    Parameters
    ----------
    ds : xr.Dataset or xr.DataArray

    Returns
    -------
    xr.Dataset or xr.DataArray
        Dataset or DataArray with dense data.

    """
    func = lambda data: data.todense() if isinstance(data, COO) else data
    return xr.apply_ufunc(func, ds)

```

**Comment by FabianHofmann** [comment]
@euronion could you point me to the parts where `atlite` uses the flattened sparse arrays you talked about?

**Comment by euronion** [comment]
Thanks @FabianHofmann for pointing this out, that sounds really promising. Especially as `sparse` is from the pydata community aiming to replace `scipy.sparse`.

The lines in question are basically the operations around the `layoutmatrix` and aggregation method in https://github.com/PyPSA/atlite/blob/7d60f68e9043cecd7f65c77b7e0f480c1f22d4b0/atlite/convert.py#L53 .

**Comment by coroa** [comment]
As long as

> It lacks layouts that are not easily generalized like CSR/CSC and depends on scipy.sparse for some computations.

is true, it will not help our use case which derives its speed from CSR-based matrix multiplication.

**Comment by euronion** [comment]
Conversion to CSR is simple `COO(...).tocsr()`, as `sparse` is `scipy.sparse` compatible. I am trying a few things over at PyPSA-EUR with the `xarray`  sparse support, significantly improving code readability.

**Comment by euronion** [comment]
So a quick verdict based on the status quo (`sparse` only supporting COO-format and many matrix operations not yet fully supported between non-sparse and sparse `xarray` DataArrays):

* Suitable for use in scripts, like in [PyPSA-EUR](https://github.com/PyPSA/pypsa-eur/blob/e8513b028a048828990bf546e784232878ee1a7b/scripts/build_renewable_profiles.py)
* Not suitable in general for `atlite`

It boils down to being usable where only small sparse arrays (<= ~ 1Mio. NNZs) or few sparse-dense operations are involved. The performance impact is measureable but acceptable. For time-series aggregation in `atlite` I did a quick test on a common PyPSA-EUR system (189x157 grid, ca. 3000 buses/regions):

* Necessary operations for bringing the time-series data into the right format (conversion to sparse array or index-stacking and reordering): Using `sparse` ca. 4x slower (~15-20s for `sparse` / one weather year)
* Aggregation (matrix broadcasting and multiplication): `sparse` ca. 200x slower (~5-10 minutes for `sparse` / one weather year)

Timings did vary since I didn't do a really clean performance test (- didn't consider it necessary).

Also the memory overhead was noticable due to manual conversion becoming necessary between non-sparse and sparse `xarray`s.

I did not test the "dirty" option of converting to `scipy.sparse` arrays. I don't think it's worth investing time at the moment but rather wait till `sparse` get's support for more formats. That should be the next big feature according to their roadmap.

@nworbmot : The numbers you were interested in earlier.
